File: src/55.ts
// —————————————————————————————————————————————————————————————————————————————
// Solution

function canJump(v:number[]) {
   let i = 0
   for (let d=0; i<v.length && i<=d; i++) 
      d = Math.max(d, i + v[i])
   return i === v.length
}

function canJump1(v:number[]) {
   let d = 0
   for (let i=0; i<v.length; i++) {
      if (i > d) return false
      d = Math.max(d, i + v[i])
   }
   return true
}

File: src/bin/55.rs
// —————————————————————————————————————————————————————————————————————————————
// Solution

// fn can_jump(nums: Vec<i32>) -> bool {
//    let mut d = 0;
//    for i in 0..nums.len() {
//       if i > d { return false }
//       d = d.max(i + nums[i] as usize);
//    }
//    return true;
// }

fn can_jump(nums: Vec<i32>) -> bool {
   let mut i = 0;
   let mut d = 0;
   while i < nums.len() && i <= d {
      d = d.max(i + nums[i] as usize);
      i += 1;
   }
   return i == nums.len();
}

fn main() {
   println!("Can jump: {}", can_jump(vec![2, 3, 1, 1, 4]));
   println!("Can jump: {}", can_jump(vec![3, 2, 1, 0, 4]));
}